src/net.cpp

```diff
@@ -2187,6 +2188,45 @@ void CConnman::ThreadMessageHandler()
     }
 }
 
+void CConnman::ThreadI2PAcceptIncoming()
+{
+    static constexpr auto err_wait_begin = 1s;
+    static constexpr auto err_wait_cap = 5min;
+    auto err_wait = err_wait_begin;
+
+    bool advertising_listen_addr = false;
+    i2p::Connection conn;
+
+    while (!interruptNet) {
+
+        if (!m_i2p_sam_session->Listen(conn)) {
+            if (advertising_listen_addr && conn.me.IsValid()) {
+                RemoveLocal(conn.me);
+                advertising_listen_addr = false;
+            }
+
+            interruptNet.sleep_for(err_wait);
+            if (err_wait < err_wait_cap) {
+                err_wait *= 2;
+            }
+
+            continue;
+        }
+
+        if (!advertising_listen_addr) {
+            AddLocal(conn.me, LOCAL_BIND);
```

Comment:
I don't entirely understand how these enums (`LOCAL_BIND`, `LOCAL_MANUAL`, etc) are meant to be used, but I wonder if this should be `LOCAL_MANUAL` instead?  That seems to be what we use for listening on an onion address: https://github.com/bitcoin/bitcoin/blob/0ca8b7e7ecd5bc537fbc1e372f6755a34a136f7f/src/torcontrol.cpp#L354

I noticed that when I'm running with `-externalip`, I don't get my i2p address listed as a local address (presumably because `fDiscover` gets defaulted to false and then anything less than `LOCAL_MANUAL` is ignored in `AddLocal()`), even though my onion address is unaffected by the setting.  Changing this to `LOCAL_MANUAL` seems to fix this issue.

Reply:
I played with the `LOCAL_*` enums, but I don't remember why I chose `LOCAL_BIND` :-( I wonder why `AddLocal()` induced by `-externalip` did not work:

https://github.com/bitcoin/bitcoin/blob/0ca8b7e7ecd5bc537fbc1e372f6755a34a136f7f/src/init.cpp#L1304-L1307

Reply:
Sorry I should have been more clear -- I was using `-externalip` with an ipv4 address (this node is listening on an ipv4 address, onion address, and i2p address).  However I noticed that with `-externalip` enabled, the `AddLocal` for the i2p address didn't work.

Reply:
> Presumably if I set `-discover=1` explicitly then it would have worked?

Yes.

Fixed in https://github.com/bitcoin/bitcoin/pull/21914.